File: Payload_Type/merlin/mythic/agent_functions/memory.py

from merlin import MerlinJob
from mythic_payloadtype_container.MythicCommandBase import *
from mythic_payloadtype_container.MythicRPC import *
import json

# Set to enable debug output to Mythic
debug = False


class MemoryArguments(TaskArguments):
    def __init__(self, command_line, **kwargs):
        super().__init__(command_line, **kwargs)
        self.args = [
            CommandParameter(
                name="method",
                cli_name="method",
                display_name="Method",
                type=ParameterType.ChooseOne,
                choices=['patch', 'read', 'write'],
                default_value='patch',
                description="The method of interaction with the agent's virtual memory",
                parameter_group_info=[ParameterGroupInfo(
                    group_name="Default",
                    ui_position=0,
                    required=True,
                )],
            ),
            CommandParameter(
                name="module",
                cli_name="module",
                display_name="Module",
                type=ParameterType.String,
                description="The module (e.g., ntdll.dll) that contains the function you want to interact with",
                default_value="ntdll.dll",
                value="ntdll.dll",
                parameter_group_info=[
                    ParameterGroupInfo(
                        group_name="Default",
                        ui_position=1,
                        required=True,
                    ),
                    ParameterGroupInfo(
                        group_name="Patch",
                        ui_position=1,
                        required=True,
                    ),
                    ParameterGroupInfo(
                        group_name="Read",
                        ui_position=1,
                        required=True,
                    ),
                    ParameterGroupInfo(
                        group_name="Write",
                        ui_position=1,
                        required=True,
                    ),
                ],
            ),
            CommandParameter(
                name="proc",
                cli_name="proc",
                display_name="Procedure",
                type=ParameterType.String,
                description="The procedure, or function, name (e.g., EtwEventWrite) that you want to interact with",
                default_value="EtwEventWrite",
                value="EtwEventWrite",
                parameter_group_info=[
                    ParameterGroupInfo(
                        group_name="Default",
                        ui_position=2,
                        required=True,
                    ),
                    ParameterGroupInfo(
                        group_name="Patch",
                        ui_position=2,
                        required=True,
                    ),
                    ParameterGroupInfo(
                        group_name="Read",
                        ui_position=2,
                        required=True,
                    ),
                    ParameterGroupInfo(
                        group_name="Write",
                        ui_position=2,
                        required=True,
                    ),
                ],
            ),
            CommandParameter(
                name="bytes",
                cli_name="bytes",
                display_name="Bytes",
                type=ParameterType.String,
                description="The bytes, as a hex string, that you want to be written to memory",
                default_value="9090C3",
                value="9090C3",
                parameter_group_info=[
                    ParameterGroupInfo(
                        group_name="Default",
                        ui_position=3,
                        required=False,
                    ),
                    ParameterGroupInfo(
                        group_name="Patch",
                        ui_position=3,
                        required=False,
                    ),
                    ParameterGroupInfo(
                        group_name="Write",
                        ui_position=3,
                        required=False,
                    ),
                ],
            ),
            CommandParameter(
                name="length",
                cli_name="length",
                display_name="Length",
                type=ParameterType.Number,
                description="The number of bytes to read from the target procedure/function",
                default_value=6,
                value=6,
                parameter_group_info=[
                    ParameterGroupInfo(
                        group_name="Default",
                        ui_position=4,
                        required=False,
                    ),
                    ParameterGroupInfo(
                        group_name="Read",
                        ui_position=4,
                        required=False,
                    ),
                ],
            ),
        ]

    async def parse_arguments(self):
        if len(self.command_line) > 0:
            if self.command_line[0] == '{':
                self.load_args_from_json_string(self.command_line)
            else:
                pass


class MemoryCommand(CommandBase):
    cmd = "memory"
    needs_admin = False
    help_cmd = "memory"
    description = "Read/Write the agent's virtual memory for the provided module and function\n" \
                  "\t - Use the \"Patch\" parameter group to read and then overwrite the target function's memory\n" \
                  "\t - Use the \"Read\" parameter group to read target function's memory\n" \
                  "\t - Use the \"Write\" parameter group to overwrite target function's memory with provided bytes\n"
    version = 1
    author = "@Ne0nd0g"
    argument_class = MemoryArguments
    attackmapping = ["T1562.001"]

    async def create_tasking(self, task: MythicTask) -> MythicTask:
        if debug:
            await MythicRPC().execute(function_name="create_output", task_id=task.id, output=f'\n[DEBUG]Input task:{task}')
            await MythicRPC().execute(function_name="create_output", task_id=task.id, output=f'\n[DEBUG]Parameter Group:{task.args.get_parameter_group_name()}')

        # Arguments
        # 0 - Method
        # 1 - Module
        # 2 - Procedure/Function
        # 3 - Bytes or Length (both as string)

        if task.args.get_parameter_group_name() == "Default":
            if task.args.get_arg("method").lower() == "read":
                args = [
                    task.args.get_arg("method"),
                    task.args.get_arg("module"),
                    task.args.get_arg("proc"),
                    str(task.args.get_arg("length")),
                ]
            else:
                args = [
                    task.args.get_arg("method").lower(),
                    task.args.get_arg("module"),
                    task.args.get_arg("proc"),
                    task.args.get_arg("bytes"),
                ]
        else:
            if task.args.get_arg("method").lower() == "read":
                args = [
                    task.args.get_parameter_group_name().lower(),
                    task.args.get_arg("module"),
                    task.args.get_arg("proc"),
                    str(task.args.get_arg("length")),
                ]
            else:
                args = [
                    task.args.get_parameter_group_name().lower(),
                    task.args.get_arg("module"),
                    task.args.get_arg("proc"),
                    task.args.get_arg("bytes"),
                ]

        # Merlin jobs.Command message type
        command = {
            "command": self.cmd,
            "args": args,
        }

        display = ""
        for arg in args:
            display += f'{arg} '
        task.display_params = f'{display}'

        task.args.add_arg("type", MerlinJob.MODULE, ParameterType.Number)
        task.args.add_arg("payload", json.dumps(command), ParameterType.String)
        task.args.remove_arg("method")
        task.args.remove_arg("module")
        task.args.remove_arg("proc")
        task.args.remove_arg("bytes")
        task.args.remove_arg("length")

        if debug:
            await MythicRPC().execute("create_output", task_id=task.id, output=f'[DEBUG]Returned task:\r\n{task}\r\n')

        return task

    async def process_response(self, response: AgentResponse):
        pass

File: Payload_Type/merlin/mythic/container/commands/memory.go
/*
Merlin is a post-exploitation command and control framework.

This file is part of Merlin.
Copyright (C) 2023  Russel Van Tuyl

Merlin is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as
published by the Free Software Foundation, either version 3 of the License, or any later version.

Merlin is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with Merlin.  If not, see <http://www.gnu.org/licenses/>.
*/

package commands

import (
	// Standard
	"fmt"
	"strings"

	// Mythic
	structs "github.com/MythicMeta/MythicContainer/agent_structs"
	"github.com/MythicMeta/MythicContainer/logging"

	// Merlin
	"github.com/Ne0nd0g/merlin/pkg/jobs"
)

// memory creates and return a Mythic Command structure that is registered with the Mythic server
func memory() structs.Command {
	attr := structs.CommandAttribute{
		SupportedOS: []string{structs.SUPPORTED_OS_WINDOWS},
	}

	method := structs.CommandParameter{
		Name:                                    "method",
		ModalDisplayName:                        "Method",
		CLIName:                                 "method",
		ParameterType:                           structs.COMMAND_PARAMETER_TYPE_CHOOSE_ONE,
		Description:                             "The method of interaction with the agent's virtual memory",
		Choices:                                 []string{"patch", "read", "write"},
		DefaultValue:                            "patch",
		SupportedAgents:                         nil,
		SupportedAgentBuildParameters:           nil,
		ChoicesAreAllCommands:                   false,
		ChoicesAreLoadedCommands:                false,
		FilterCommandChoicesByCommandAttributes: nil,
		DynamicQueryFunction:                    nil,
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   true,
				GroupName:             "Default",
				UIModalPosition:       0,
				AdditionalInformation: nil,
			},
		},
	}

	module := structs.CommandParameter{
		Name:                                    "module",
		ModalDisplayName:                        "Module",
		CLIName:                                 "module",
		ParameterType:                           structs.COMMAND_PARAMETER_TYPE_STRING,
		Description:                             "The module (e.g., ntdll.dll) that contains the function you want to interact with",
		Choices:                                 nil,
		DefaultValue:                            "ntdll.dll",
		SupportedAgents:                         nil,
		SupportedAgentBuildParameters:           nil,
		ChoicesAreAllCommands:                   false,
		ChoicesAreLoadedCommands:                false,
		FilterCommandChoicesByCommandAttributes: nil,
		DynamicQueryFunction:                    nil,
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   true,
				GroupName:             "Default",
				UIModalPosition:       1,
				AdditionalInformation: nil,
			},
			{
				ParameterIsRequired:   true,
				GroupName:             "Patch",
				UIModalPosition:       1,
				AdditionalInformation: nil,
			},
			{
				ParameterIsRequired:   true,
				GroupName:             "Read",
				UIModalPosition:       1,
				AdditionalInformation: nil,
			},
			{
				ParameterIsRequired:   true,
				GroupName:             "Write",
				UIModalPosition:       1,
				AdditionalInformation: nil,
			},
		},
	}

	proc := structs.CommandParameter{
		Name:                                    "proc",
		ModalDisplayName:                        "Procedure",
		CLIName:                                 "proc",
		ParameterType:                           structs.COMMAND_PARAMETER_TYPE_STRING,
		Description:                             "The procedure, or function, name (e.g., EtwEventWrite) that you want to interact with",
		Choices:                                 nil,
		DefaultValue:                            "EtwEventWrite",
		SupportedAgents:                         nil,
		SupportedAgentBuildParameters:           nil,
		ChoicesAreAllCommands:                   false,
		ChoicesAreLoadedCommands:                false,
		FilterCommandChoicesByCommandAttributes: nil,
		DynamicQueryFunction:                    nil,
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   true,
				GroupName:             "Default",
				UIModalPosition:       2,
				AdditionalInformation: nil,
			},
			{
				ParameterIsRequired:   true,
				GroupName:             "Patch",
				UIModalPosition:       2,
				AdditionalInformation: nil,
			},
			{
				ParameterIsRequired:   true,
				GroupName:             "Read",
				UIModalPosition:       2,
				AdditionalInformation: nil,
			},
			{
				ParameterIsRequired:   true,
				GroupName:             "Write",
				UIModalPosition:       2,
				AdditionalInformation: nil,
			},
		},
	}

	bytes := structs.CommandParameter{
		Name:                                    "bytes",
		ModalDisplayName:                        "Bytes",
		CLIName:                                 "bytes",
		ParameterType:                           structs.COMMAND_PARAMETER_TYPE_STRING,
		Description:                             "The bytes, as a hex string, that you want to be written to memory",
		Choices:                                 nil,
		DefaultValue:                            "9090C3",
		SupportedAgents:                         nil,
		SupportedAgentBuildParameters:           nil,
		ChoicesAreAllCommands:                   false,
		ChoicesAreLoadedCommands:                false,
		FilterCommandChoicesByCommandAttributes: nil,
		DynamicQueryFunction:                    nil,
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   false,
				GroupName:             "Default",
				UIModalPosition:       3,
				AdditionalInformation: nil,
			},
			{
				ParameterIsRequired:   true,
				GroupName:             "Write",
				UIModalPosition:       3,
				AdditionalInformation: nil,
			},
		},
	}

	// Needed a parameter with a different name, so Mythic doesn't get confused on the exclusive set
	patchBytes := structs.CommandParameter{
		Name:                                    "patch-bytes",
		ModalDisplayName:                        "Bytes",
		CLIName:                                 "patch-bytes",
		ParameterType:                           structs.COMMAND_PARAMETER_TYPE_STRING,
		Description:                             "The bytes, as a hex string, that you want to be written to memory. Only used with the 'patch' method.",
		Choices:                                 nil,
		DefaultValue:                            "9090C3",
		SupportedAgents:                         nil,
		SupportedAgentBuildParameters:           nil,
		ChoicesAreAllCommands:                   false,
		ChoicesAreLoadedCommands:                false,
		FilterCommandChoicesByCommandAttributes: nil,
		DynamicQueryFunction:                    nil,
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   true,
				GroupName:             "Patch",
				UIModalPosition:       3,
				AdditionalInformation: nil,
			},
		},
	}

	length := structs.CommandParameter{
		Name:                                    "length",
		ModalDisplayName:                        "Length",
		CLIName:                                 "length",
		ParameterType:                           structs.COMMAND_PARAMETER_TYPE_NUMBER,
		Description:                             "The number of bytes to read from the target procedure/function",
		Choices:                                 nil,
		DefaultValue:                            6,
		SupportedAgents:                         nil,
		SupportedAgentBuildParameters:           nil,
		ChoicesAreAllCommands:                   false,
		ChoicesAreLoadedCommands:                false,
		FilterCommandChoicesByCommandAttributes: nil,
		DynamicQueryFunction:                    nil,
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   false,
				GroupName:             "Default",
				UIModalPosition:       4,
				AdditionalInformation: nil,
			},
			{
				ParameterIsRequired:   true,
				GroupName:             "Read",
				UIModalPosition:       4,
				AdditionalInformation: nil,
			},
		},
	}

	params := []structs.CommandParameter{method, module, proc, bytes, patchBytes, length}
	command := structs.Command{
		Name:                  "memory",
		NeedsAdminPermissions: false,
		HelpString:            "memory <method>",
		Description: "Read/Write the agent's virtual memory for the provided module and function." +
			" Use the \"Patch\" parameter group to read and then overwrite the target function's memory. " +
			"Use the \"Read\" parameter group to read target function's memory. " +
			"Use the \"Write\" parameter group to overwrite target function's memory with provided bytes.",
		Version:                        0,
		SupportedUIFeatures:            nil,
		Author:                         "@Ne0nd0g",
		MitreAttackMappings:            []string{"T1562.001"},
		ScriptOnlyCommand:              false,
		CommandAttributes:              attr,
		CommandParameters:              params,
		AssociatedBrowserScript:        nil,
		TaskFunctionOPSECPre:           nil,
		TaskFunctionCreateTasking:      memoryCreateTask,
		TaskFunctionProcessResponse:    nil,
		TaskFunctionOPSECPost:          nil,
		TaskFunctionParseArgString:     taskFunctionParseArgString,
		TaskFunctionParseArgDictionary: taskFunctionParseArgDictionary,
		TaskCompletionFunctions:        nil,
	}

	return command
}

// memoryCreateTask takes a Mythic Task and converts into a Merlin Job that is encoded into JSON and subsequently sent to the Merlin Agent
func memoryCreateTask(task *structs.PTTaskMessageAllData) (resp structs.PTTaskCreateTaskingMessageResponse) {
	pkg := "mythic/container/commands/memory/memoryCreateTask()"
	resp.TaskID = task.Task.ID

	method, err := task.Args.GetStringArg("method")
	if err != nil {
		err = fmt.Errorf("%s: there was an error getting the \"method\" argument: %s", pkg, err)
		resp.Error = err.Error()
		resp.Success = false
		logging.LogError(err, "returning with error")
		return
	}

	module, err := task.Args.GetStringArg("module")
	if err != nil {
		err = fmt.Errorf("%s: there was an error getting the \"module\" argument: %s", pkg, err)
		resp.Error = err.Error()
		resp.Success = false
		logging.LogError(err, "returning with error")
		return
	}

	proc, err := task.Args.GetStringArg("proc")
	if err != nil {
		err = fmt.Errorf("%s: there was an error getting the \"proc\" argument: %s", pkg, err)
		resp.Error = err.Error()
		resp.Success = false
		logging.LogError(err, "returning with error")
		return
	}

	bytes, err := task.Args.GetStringArg("bytes")
	if err != nil {
		err = fmt.Errorf("%s: there was an error getting the \"bytes\" argument: %s", pkg, err)
		resp.Error = err.Error()
		resp.Success = false
		logging.LogError(err, "returning with error")
		return
	}

	patchBytes, err := task.Args.GetStringArg("patch-bytes")
	if err != nil {
		err = fmt.Errorf("%s: there was an error getting the 'patch-bytes' argument: %s", pkg, err)
		resp.Error = err.Error()
		resp.Success = false
		logging.LogError(err, "returning with error")
		return
	}

	length, err := task.Args.GetNumberArg("length")
	if err != nil {
		err = fmt.Errorf("%s: there was an error getting the \"length\" argument: %s", pkg, err)
		resp.Error = err.Error()
		resp.Success = false
		logging.LogError(err, "returning with error")
		return
	}

	job := jobs.Command{
		Command: "memory",
	}

	var disp string
	switch strings.ToLower(task.Task.ParameterGroupName) {
	case "default":
		switch strings.ToLower(method) {
		case "read":
			job.Args = []string{method, module, proc, fmt.Sprintf("%d", int(length))}
			disp = fmt.Sprintf("%s %s:%s %d", method, module, proc, int(length))
		case "patch":
			job.Args = []string{method, module, proc, bytes}
			disp = fmt.Sprintf("%s %s:%s %s", method, module, proc, bytes)
		case "write":
			job.Args = []string{method, module, proc, bytes}
			disp = fmt.Sprintf("%s %s:%s %s", method, module, proc, bytes)
		default:
			err = fmt.Errorf("%s: unhandled parameter group name %s", pkg, task.Task.ParameterGroupName)
			resp.Error = err.Error()
			resp.Success = false
			logging.LogError(err, "returning with error")
			return
		}
	case "read":
		job.Args = []string{"read", module, proc, fmt.Sprintf("%d", int(length))}
		disp = fmt.Sprintf("%s %s:%s %d", strings.ToLower(task.Task.ParameterGroupName), module, proc, int(length))
	case "patch":
		job.Args = []string{"patch", module, proc, patchBytes}
		disp = fmt.Sprintf("%s %s:%s %s", strings.ToLower(task.Task.ParameterGroupName), module, proc, patchBytes)
	case "write":
		job.Args = []string{"write", module, proc, bytes}
		disp = fmt.Sprintf("%s %s:%s %s", strings.ToLower(task.Task.ParameterGroupName), module, proc, bytes)
	}

	mythicJob, err := ConvertMerlinJobToMythicTask(job, jobs.MODULE)
	if err != nil {
		err = fmt.Errorf("%s: there was an error converting the Merlin job to a Mythic job: %s", pkg, err)
		resp.Error = err.Error()
		resp.Success = false
		logging.LogError(err, "returning with error")
		return
	}

	task.Args.SetManualArgs(mythicJob)

	resp.DisplayParams = &disp
	resp.Success = true

	return
}
